packages/server/lib/plugins/index.js

```diff
@@ -61,7 +61,11 @@ const getChildOptions = (config) => {
 
   // To be removed on update to webpack >= 5.61, which no longer relies on
   // node's builtin crypto.hash function.
-  if (semver.satisfies(config.resolvedNodeVersion, '>=17.0.0')) {
+  // Node 17+ claims to support the new --openssl-legacy-provider option
+  // unconditionally, but it doesn't exist if Node was built against OpenSSL
+  // version 1.x.
+  if (semver.satisfies(config.resolvedNodeVersion, '>=17.0.0') &&
+      !process.versions.openssl.startsWith("1")) {
```

Comment:
Given this issue is caused by Cypress's usage of wepback 4 + Node 17,  the "best" solution for this issue would be to upgrade to webpack 5. I am looking into that work, however, for a (hopefully temporary) go-forward I think you change is on the right track since webpack uses the `md4` cyrpto algorithm internally.

I am wonder if we need to make this solution a bit more robust though. The original change here assumed default the OpenSSL Library provided in Node 17 to be the OpenSSL 3.0.0 which introduced the `--openssl-legacy-provider` flag. Your change is assuming assuming Node 17 + OpenSSL 1.x.

In your summary / linked issues to Cloudflare, you noted that Node can be built with the `--shared-openssl` build parameter. This lead me to look further into Node's build options & their[build documentation](https://github.com/nodejs/node/blob/master/BUILDING.md#building-nodejs-with-fips-compliant-openssl). It looks like the OpenSSL library / configuration( & the cyrpto library) could be pretty open-ended if someone wanted to build node with custom paths.

Would it be possible for users to use a custom OpenSSL library that doesn't accept the `--openssl-legacy-provider` flag? Would that cause `process.versions.openssl` to be undefined?

Reply:
Also, can we update the existing tests to add a validation check for this?

Reply:
> Your change is assuming assuming Node 17 + OpenSSL 1.x.

My change does not assume Node 17 + OpenSSL 1.x, it just checks for that possibility and skips adding the option if that happens to be the case. The option is still applied on Node 17 + OpenSSL 3.x.

---

I fully agree that the best course of action for Cypress would be to upgrade to Webpack 5.54+ and changing its `output.hashFunction` option to `xxhash64` (which Webpack 6 will use by default). It's the right thing to do regardless from project health viewpoint. This whole workaround could then be removed.

Regarding uncertainty about building Node with custom versions of OpenSSL, Node.js core team says [building against any shared libraries is effectively unsupported](https://github.com/nodejs/node/issues/35213#issuecomment-695753042). I wouldn't support anything beyond OpenSSL 1.x and OpenSSL 3.x for this hopefully temporary fix.

If you wanted to make this more robust, I'd require `crypto` and check whether `crypto.constants.OPENSSL_VERSION_NUMBER > 0x30000000`. But that's essentially a more complicated way of testing the same thing. (I hope this workaround is long gone by the time OpenSSL 11 is released.)

Regarding cases where `process.versions.openssl` would be undefined, in theory it's possible to [build Node.js without SSL at all](https://github.com/nodejs/node/blob/c85e67b600c07c0c37bebd077579f102747062a4/configure.py#L671-L675). (This is obviously unsupported by most of the ecosystem and no major distro ships Node.js built like this.) Via this issue we've established that Cypress transitively relies on `crypto` being available; if you wanted to [check for that](https://nodejs.org/api/crypto.html#determining-if-crypto-support-is-unavailable), I'd do it much earlier.

Reply:
> My change does not assume Node 17 + OpenSSL 1.x, it just checks for that possibility and skips adding the option if that happens to be the case. 

To me it seems like it might be safer to check for the possibility of OpenSSL 3.x and add it if its there. @mjhenkes  thoughts?

Reply:
This switch will be needed with OpenSSL versions from now on until you upgrade Webpack. If OpenSSL were to increase their major version again checking explicitly for OpenSSL 3 would break.

Reply:
True, but there's no guarantee the next MVB of OpenSSL will support the Legacy Provider and it's corresponding `--openssl-legacy-provider` flag either. 🤷🏻‍♀️ This is good conversation btw. I appreciate your responsiveness.

Reply:
I appreciate you making sure it's the appropriate fix.

We **know** for sure that the `--openssl-legacy-provider` option is not there with OpenSSL version 1.x. I also want to emphasize that we are now _fixing existing breakage_ -- something many people have now encountered in the wild, even based on reports from the very small subset of users that engage with the project via GitHub.

We have no idea when or if OpenSSL 4 will be released -- and by the time OpenSSL 4 might be released, I really truly hope Cypress has long since moved on from webpack@4 and removed this workaround altogether!

---

Regarding support for the legacy provider, they just added it in OpenSSL 3 to draw a clear line between believed-safe and known-broken algorithms and ciphers. They had the opportunity to completely throw these algorithms away here, but they chose to keep supporting them to let their clients decide when and how to upgrade. They wisely mitigated the risk by not loading this provider by default and having their clients *explicitly* opt-in to it.

I would not expect this legacy provider to disappear anytime soon, as otherwise the currently clear upgrade path from OpenSSL 1.x (which is quite literally *everywhere*) would suffer. This is of vital importance to the OpenSSL team since there still are multiple CVEs found in OpenSSL every year -- they need to keep providing as smooth upgrade experience as possible.

(I suspect that's also the reason why many rolling release distros have chosen to build Node.js with `--shared-openssl` in the first place -- they believe they will be better off updating OpenSSL systemwide rather than relying on the Node.js team for a new build. Unfortunately these very infrequent big OpenSSL upgrades then take more time to roll out since you have to fix and rebuild large parts of the whole distro. Arch Linux is looking to upgrade at the [beginning of this year](https://lists.archlinux.org/pipermail/arch-dev-public/2021-December/030576.html).)

Reply:
I suppose one other way of solving this would be to copy the approach I took in [fixing wrangler](https://github.com/cloudflare/wrangler/pull/2172/files) by spawning `node --help` and grepping for `--openssl-legacy-provider`. It's going to be slower and much harder to test, but maybe slightly more robust.

I have no insight into your plans and schedule wrt. webpack 5 adoption, so if this workaround is going to stay here for a longer while it might be wiser to take that approach. Who knows how long Node.js will keep that option around.

Reply:
We can go head with what is here -- a new node process would be overkill. Based on your response, we know for sure `--openssl-legacy-provider` option is not supported with OpenSSL version 1.x and could potentially be supported in OpenSSL >3.x. Like you said v4 likely won't come quickly and we should be (fingers crossed) on a newer version of webpack by the time this would happen.